pkg/storage/iterator.go

```diff
@@ -415,6 +423,13 @@ func fetchLazyChunks(ctx context.Context, chunks []*chunkenc.LazyChunk) error {
 		}(fetcher, chunks)
 	}
 
+	// remove data of any invalid chunk
+	for _, fetcher := range invalidChunks {
```

Comment:
You need to remove empty chunk from the `chunks` variable. And this needs to happens when we have waited for all errors.

after:

```golang
var lastErr error
	for i := 0; i < len(chksByFetcher); i++ {
		if err := <-errChan; err != nil {
			lastErr = err
		}
	}

if lastErr != nil {
 return lastErr
}
i := 0 // output index
for _, c := range chunks {
    if c.Chunk.Data != nil {
        chunks[i] = c
        i++
    }
}
chunks = chunks[:i]
return nil
```

Reply:
Got you

Reply:
@cyriltovena I see a problem here. I may not be correct though.
From the following function https://github.com/grafana/loki/blob/master/pkg/storage/iterator.go#L280, we first partitionBySeriesChunks, which partitions chunks based on overlapping iterator. At this stage say `chunksBySeries` has chunks as below
chksBySeries after partition
```
[0xc00011c400]
[0xc00011c480]
[0xc00011c500 0xc00011c600]
[0xc00011c580]
```
Later first chunks are loaded from `loadFirstChunks` which is passed with value `chksBySeries`. After this `chksBySeries` looks as below 
```
[0xc00011c400]
[0xc00011c480]
[0xc00011c500 0xc00011c600]
[0xc00011c580]
```

and in turn `fetchLazyChunks` would have picked following chunks
`[0xc00011c400 0xc00011c480 0xc00011c500 0xc00011c580]`

Even at this point `chkBySeries` has below values
```
[0xc00011c400]
[0xc00011c480]
[0xc00011c500 0xc00011c600]
[0xc00011c580]
```
Later `filterSeriesByMatchers` is called with value `chksBySeries` and `chksBySeries` has following values
```
[[0xc00011c500 0xc00011c600][0xc00011c580]]
```
Now, say `0xc00011c600` chunk returns `invalid chunk checksum` when doing` fetchLazyChunks` and the chunks picked are `[0xc00011c500],[0xc00011c580]`, Iterator is not affected with this because `buildIterators` is passed with `chksBySeries` which iterates following chunks
`[[0xc00011c500 0xc00011c600][0xc00011c580]]`. Since, `0xc00011c600` is having no data, iterator errors saying `chunks not loaded`. 

May be we should check for `nil Data` chunk here https://github.com/grafana/loki/blob/master/pkg/storage/iterator.go#L335 and skip it 
OR
Re arrange `chkBySeries` after fetching lazy chunks before calling `buildIterators` here https://github.com/grafana/loki/blob/master/pkg/storage/iterator.go#L305

Reply:
I'm off for a week. We'll get back to you then.

Reply:
Sure. Tried to add a dummy code here https://play.golang.org/p/HFXOy5Ek2xd. This should explain what's actually happening.

Reply:
The only problem I see with my suggestions is that you might actually discard a lot of chunks if the first chunk is corrupted, but that's still better than crashing so let's do that and we can review this later if someone has an issue with it.

Reply:
`discard a lot of chunks if the first chunk is corrupted`, this is the same behaviour currently right?

Reply:
yes

Reply:
So, just leave things as is and close PR?

Reply:
No currently it crashes. I want this function to be changed like we discussed.

Reply:
Sorry I was explaining things without pushing code. The code added fails [here](https://cloud.drone.io/grafana/loki/2199/1/2). The test code is not neat. Just validating things. However, we get into below scenario.

The chunks we get here https://github.com/grafana/loki/pull/1713/files#diff-7d61465906fc121b68e860420fcd1c87R440 may be like 
`[chunk1, chunk3]`. However, `chksBySeries` here https://github.com/grafana/loki/blob/64a1c0d96bd033d67c634aff195fe1b0b96441c5/pkg/storage/iterator.go#L305 might have chunks as `[[chunk1, chunk2][chunk3]]`. So, if `chunk2` is having `nil` data due to `invalid checksum` `chksBySeries` don't drop the chunk. Which leads to `chunk is not loaded` here https://github.com/grafana/loki/blob/64a1c0d96bd033d67c634aff195fe1b0b96441c5/pkg/chunkenc/lazy_chunk.go#L29

We should re-arrange the chunks before calling `buildIterators` https://github.com/grafana/loki/blob/64a1c0d96bd033d67c634aff195fe1b0b96441c5/pkg/storage/iterator.go#L305

Reply:
Yes That's what I tried to explain you should remove the lazy chunk from the array.

Reply:
I am just checking for easy ways to do this  Removing a chunk from an array is quiet a costly operation. We would have to remove an element and shift other elements. When we have large number of chunks to load, this can turn out to be time taking process. 

One easy way is to label the chunk as `invalid` or `valid` or may be a boolean. Then in the iterators we can check if the chunk is valid or not. If invalid we can skip and continue.

what's your thought on this?

Reply:
@cyriltovena Need your feedback on this